src/rshim.c

```diff
@@ -2073,42 +2116,56 @@ int rshim_set_drop_mode(rshim_backend_t *bd, int value)
 {
   int old_value;
   int rt = 0;
+  int has_bd_lock;
+
+  has_bd_lock = !pthread_mutex_trylock(&bd->mutex);
 
-  pthread_mutex_lock(&bd->mutex);
   old_value = (int)bd->drop_mode;
   value = !!value;
   if (value == old_value) {
-    pthread_mutex_unlock(&bd->mutex);
-    return -EALREADY;
+    if (has_bd_lock)
+      pthread_mutex_unlock(&bd->mutex);
+    return 0;
   }
 
-  bd->drop_mode = 0;
-  if (bd->enable_device && bd->enable_device(bd, value ? false : true))
-    bd->drop_mode = 1;
-  else
-    bd->drop_mode = value;
```

Comment:
We can't move this block...
Considering drop mode changed from 1 to 0, it needs to enable the device first or else the following rshim_access_check() will not work.